server/quantization.go

```diff
@@ -0,0 +1,273 @@
+package server
+
+import (
+	"fmt"
+	"io"
+	"log/slog"
+	"maps"
+	"os"
+	"strings"
+	"unsafe"
+
+	fsggml "github.com/ollama/ollama/fs/ggml"
+	"github.com/ollama/ollama/ml/backend/ggml"
+)
+
+type quantizer struct {
+	*os.File
+	offset     uint64
+	from, to   *fsggml.Tensor
+	progressFn func(n uint64)
+}
+
+func (q quantizer) WriteTo(w io.Writer) (int64, error) {
+	quantize := q.from.Kind != q.to.Kind
+	sr := io.NewSectionReader(q, int64(q.offset), int64(q.from.Size()))
+	data := make([]byte, q.from.Size())
+	_, err := io.ReadFull(sr, data)
+	if err != nil {
+		slog.Warn("file read error", "tensor", q.from.Name, "file", q.Name(), "error", err)
+		return 0, fmt.Errorf("unable to read tensor %s from %s: %s", q.from.Name, q.Name(), err)
+	}
+	if quantize {
+		var f32s []float32
+		newType := fsggml.TensorType(q.to.Kind)
+		if fsggml.TensorType(q.from.Kind) == fsggml.TensorTypeF32 {
+			f32s = unsafe.Slice((*float32)(unsafe.Pointer(&data[0])), q.from.Elements())
+		} else {
+			f32s = ggml.ConvertToF32(data, q.from.Kind, q.from.Elements())
+		}
+		data = ggml.Quantize(newType, f32s, q.from.Shape)
+	}
+
+	n, err := w.Write(data)
+	q.progressFn(q.from.Size())
+	return int64(n), err
```

Comment:
Shouldn't buffer the data if `!quantize`

```suggestion
	if !quantize {
		n, err := io.Copy(w, sr)
		q.progressFn(q.from.Size())
		return int64(n), err
	}

	data := make([]byte, q.from.Size())
	_, err := io.ReadFull(sr, data)
	if err != nil {
		slog.Warn("file read error", "tensor", q.from.Name, "file", q.Name(), "error", err)
		return 0, fmt.Errorf("unable to read tensor %s from %s: %s", q.from.Name, q.Name(), err)
	}

	var f32s []float32
	newType := fsggml.TensorType(q.to.Kind)
	if fsggml.TensorType(q.from.Kind) == fsggml.TensorTypeF32 {
		f32s = unsafe.Slice((*float32)(unsafe.Pointer(&data[0])), q.from.Elements())
	} else {
		f32s = ggml.ConvertToF32(data, q.from.Kind, q.from.Elements())
	}
	data = ggml.Quantize(newType, f32s, q.from.Shape)

	n, err := w.Write(data)
	q.progressFn(n)
	return int64(n), err
```